spec.go

```diff
@@ -225,6 +225,60 @@ type SourceBuild struct {
 	Args map[string]string `yaml:"args,omitempty" json:"args,omitempty"`
 }
 
+// SourceInlineContent is used to specify the content of an inline source.
+type SourceInlineFile struct {
+	// Content is the contents.
+	Contents string `yaml:"content,omitempty" json:"content,omitempty"`
+	// Permissions is the octal file permissions to set on the file.
+	Permissions fs.FileMode `yaml:"permissions,omitempty" json:"permissions,omitempty"`
+	// UID is the user ID to set on the directory and all files and directories within it.
+	// UID must be greater than 0
+	UID int `yaml:"uid,omitempty" json:"uid,omitempty"`
+	// GID is the group ID to set on the directory and all files and directories within it.
+	// GID must be greater than 0
+	GID int `yaml:"gid,omitempty" json:"gid,omitempty"`
+}
+
+// SourceInlineDir is used by by [SourceInline] to represent a filesystem directory.
+type SourceInlineDir struct {
+	// Files is the list of files to include in the directory.
+	// The map key is the name of the file.
+	//
+	// Files with path separators in the key will be rejected.
+	//
+	// Files are created after directories are created.
+	Files map[string]*SourceInlineFile `yaml:"files,omitempty" json:"files,omitempty"`
+	// Dirs is the list of directories to include in the directory.
+	// The map key is the name of the directory.
+	//
+	// Directories with path separators in the key will be rejected.
+	//
+	// Directories are created before files are created.
+	Dirs map[string]*SourceInlineDir `yaml:"dirs,omitempty" json:"dirs,omitempty"`
+	// Permissions is the octal permissions to set on the directory.
+	Permissions fs.FileMode `yaml:"permissions,omitempty" json:"permissions,omitempty"`
+
+	// UID is the user ID to set on the directory and all files and directories within it.
+	// UID must be greater than 0
+	UID int `yaml:"uid,omitempty" json:"uid,omitempty"`
+	// GID is the group ID to set on the directory and all files and directories within it.
+	// GID must be greater than 0
+	GID int `yaml:"gid,omitempty" json:"gid,omitempty"`
+}
```

Comment:
Inline files make a lot of sense for one-off things that don't deserve their own file, but I don't know if I would want to allow directories. It seems unnecessary to allow complex nested inlining when the filesystem is right there and available as a `SourceTypeContext`.

I think it could get quickly get very indented and cluttered. This functionality is already available by other means, for example a `Build` source type with a local dir provided as a context and a small inline dockerfile to set the perms and uid/gid of the files/directories, which can be more concise due to `(chown|chmod) -R`.

Reply:
Context is not really the same thing.
Context is client-side, different clients have different contexts.

Using `Build` is also significantly more complex (on the backend) and requires a remote image to do anything at all.

I thought I put this in the commit but apparently I did not: The main intent for these is to simplify testing and not require tests to hit registries.

Reply:
> Context is not really the same thing. Context is client-side, different clients have different contexts.
> 
Understood

> Using `Build` is also significantly more complex (on the backend) and requires a remote image to do anything at all.
> 
Make sense

> I thought I put this in the commit but apparently I did not: The main intent for these is to simplify testing and not require tests to hit registries.
Sounds good. Was wondering what the reason was for this.

Reply:
> Context is not really the same thing. Context is client-side, different clients have different contexts.
> 

So actually I'm not sure I fully understand this. You mean that a context can be all kinds of things, depending on the client? For example, with `buildctl` I can do this:

```bash
buildctl build \                                                                                       --frontend=dockerfile.v0 \                                                                     --local context=. \                                                                            --oci-layout=foo=./oci \                                                                       --opt context:bar=oci-layout:foo@sha256:760c7b6<...> \                                         --local dockerfile=. \
        --output=type=local,dest=../hello

# Dockerfile
FROM bar
```

An in this case, the directory `oci` is not actually the context (because it's json blobs / tarballs of the oci layout dir), the actual context is the fs of the image *represented* by the `oci` layout dir.

Is that right?

Reply:
I mean a context is dependent on the build invocation.
The client tells you what the context is and is not declarative.